### ISSUE TITLE
Tizen 6.5 M2 Items

### ISSUE THREAD

**Issue by zhoonit** [nnstreamer/nntrainer#1033]
This issue is to keep track items for tizen 6.5 m2

scheduled issue: 📚
scheduled pr: ♻️ 

## Major API Updates
1. Add save states, read states api 
   -  `ml_train_model_save_states`, `ml_train_model_read_states` and ccapi respectively.
    #1345(proposal) ♻️ implementation ♻️ test
   - introduce new state saving format ~(using flatbuffer or directly using tflite format)~
1. Refactor data handle
   - Setting train, validation, test independently
   #1371, #1381, #1383
   - tct, remove label.dat
1. Add way to get input shape, out put shape (number of classes)
   - `ml_train_model_get_input_tensors_info`, `ml_train_model_get_output_tensors_info` and ccapi respectively 
   - ~probably~ going to use `ml_tensors_` from nnstreamer side.
   #1346, #1430
1. enhance summary api (expose weight with this) **delayed**
1. Add optimizer section in ini  #1032
1. Layers
   - Op resolver  #659 (probably not going to do this, if not do this, only gonna support nnstreamer_backbone)
   - Add enum types
      - backbone
      - centroid_knn
      - l2norm
      - centering (need discussion, imo it's not needed)
      - Possibly some optimization techniques (cc. @lhs8928 )
1. deprecate old dataset properties

## Internals
1. Add dynamic var_grad support.
   - has a default tensor dimensions, returns unallocated tensor, 
1. Integrate simpleshot
1. rework centroid knn
   - to make them have dynamic weight
   - to make them have dynamic output
1. tensor filter
   - exploit or always use flextype if output is dynamic, need consultation with @jaeyun-jung before working
1. dataset 
   - delete data_label and label.dat (#1376)
   - data_test is not used anywhere, please fix
   - dataset to get multi input, label with specifeid label shape (#1374)
   - we will have a three copy of dataset for now. (#1379)

## Misc
- ~Check if non-one-hot encoded label is fully eligible --> this is elegible~
- Retouch documentation about dataset
- remove setProperty("user_data") from the documentation
- Fix typo in the documentation example (ml_train_set_property)

**Comment by zhoonit** [comment]
- Support backbone as a standalone layer (TPL requirement)

**Comment by daeinki** [comment]
One issue on  SimpleShot learning of NNTrainer, which is  what we discussed already in online, :)

SimpleShot learning of NNTrainer doesn't allow a new data to be added to existing weight table in runtime because the weight table has a fixed size - dataset consists of data and one-hot encoding table as one pair according to  "Set a dataset from a file" section of relevant document[1],
```
[float array for data1][label(on-hot encoded) for data1][float array][label]...
```

And seems that the use of one-hot encoding table for label of each data makes the size of the weight table to be fixed.

[1] https://docs.tizen.org/application/native/guides/machine-learning/machine-learning-training/

**Comment by zhoonit** [comment]
Summarizing offline discussion with @daeinki.

Below item should be done slightly before M2 (by mid-july)

- Flexible CentroidKNN (increasing class numbers)
- Number encoded label (It is already supported, but double check if it's okay with databuffer)
- ...

**Comment by daeinki** [comment]
One request,

Could you add a new API which provides meta information to the model file which contains weight values?
There is a use case that user wants to know below information from the model file,
- width and height values to each data.
- the number of classes contained in the model file.
- and so on.

In case of SimpleShot learning, there is no way for user can know such information from the model file.

**Comment by zhoonit** [comment]
Thank you for sharing :)

If I enhance below api for the simpleshot, would it be working for you?

<details> <summary>Tizen API</summary>

```cpp
/**
 * @brief Gets the summary of the neural network model.
 * @details Use this function to get the summary of the neural network model.
 * @since_tizen 6.0
 * @remarks If the function succeeds, @a summary should be released using
 * free().
 * @param[in] model The NNTrainer model handle to get summary.
 * @param[in] verbosity Verbose level of the summary
 * @param[out] summary The summary of the current model. Avoid logic to parse
 * and exploit @a summary if possible.
 * @return @c 0 on success. Otherwise a negative error value.
 * @retval #ML_ERROR_NONE Successful.
 * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
 * @retval #ML_ERROR_INVALID_PARAMETER Invalid parameter.
 */
int ml_train_model_get_summary(ml_train_model_h model,
                               ml_train_summary_type_e verbosity,
                               char **summary);

/**
 * @brief Enumeration for the neural network summary verbosity of NNTrainer.
 * @since_tizen 6.0
 */
typedef enum {
  ML_TRAIN_SUMMARY_MODEL = 0, /**< Overview of model
                                   summary with one-line layer information */
  ML_TRAIN_SUMMARY_LAYER, /**< Detailed model summary with layer properties */
  ML_TRAIN_SUMMARY_TENSOR /**< Model summary layer's including weight
                             information */
} ml_train_summary_type_e;
```
</details>

**Comment by daeinki** [comment]
> Thank you for sharing :)
> 
> If I enhance below api for the simpleshot, would it be working for you?

Yeah, it would work I think. :)

> 
> Tizen API
> ```c++
> /**
>  * @brief Gets the summary of the neural network model.
>  * @details Use this function to get the summary of the neural network model.
>  * @since_tizen 6.0
>  * @remarks If the function succeeds, @a summary should be released using
>  * free().
>  * @param[in] model The NNTrainer model handle to get summary.
>  * @param[in] verbosity Verbose level of the summary
>  * @param[out] summary The summary of the current model. Avoid logic to parse
>  * and exploit @a summary if possible.
>  * @return @c 0 on success. Otherwise a negative error value.
>  * @retval #ML_ERROR_NONE Successful.
>  * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
>  * @retval #ML_ERROR_INVALID_PARAMETER Invalid parameter.
>  */
> int ml_train_model_get_summary(ml_train_model_h model,
>                                ml_train_summary_type_e verbosity,
>                                char **summary);
> 
> /**
>  * @brief Enumeration for the neural network summary verbosity of NNTrainer.
>  * @since_tizen 6.0
>  */
> typedef enum {
>   ML_TRAIN_SUMMARY_MODEL = 0, /**< Overview of model
>                                    summary with one-line layer information */
>   ML_TRAIN_SUMMARY_LAYER, /**< Detailed model summary with layer properties */
>   ML_TRAIN_SUMMARY_TENSOR /**< Model summary layer's including weight
>                              information */
> } ml_train_summary_type_e;
> ```

**Comment by myungjoo** [comment]
@daeinki @zhoonit : This API will be Tizen 6.5 M2 public API; thus, you need to be certain that this API is in a good form and worth guaranteeing backward compatibility. 

@zhoonit @kparichay @jijoongmoon : Please keep in mind that you will need to start ACR process around July. (along with sample apps, documents, and test cases). Besides, you need to start thinking about .NET API set.

**Comment by zhoonit** [comment]
Make some part of simpleshot to be part of the NNTrainer

**Comment by jijoongmoon** [comment]
> (using flatbuffer or directly using tflite format)

I'm not sure we can support tflite format in time. Also I don't think we need to support it right now. Opening API means it should work for the all the layers we provides. ( I guess about 10 or so.. ). Moreover, we need to support our format for all the layers we have first using ini_interpreter.

> Add way to get input shape, out put shape (number of classes)

I think it is not difficult to support this API. but using ```ml_tensors_``` needs more discussion. I cannot see clear reason for this. 
It is info.. not the data of tensors. let me know if I'm wrong. If it is not, please share with me.

>enhance summary API 

This is good. but what it means about the weight? Is it something like summary of network like keras or tensorflow or something else. Please elaborate more.

> Add optimizer section in ini 

which is already done I guess.

> centroid_knn

I think we need to discuss more about how we support this knn. as an network type or as an layer. First we look at how the other framework handle and discuss more.
For others like l2norm, I guess this can be formal layer in nntrainer, not custom layer. 

> dynamic weight & dynamic output

I guess we are not going to change the dimension of weight and output at runtime. I guess, from the Application point of view, the user will add additional class and then we could initialize network again and set the dimension properly. I assume it is not so hard to support this feature.

> tensor filter 

I think if the tensor filter is in pipeline, then the pipeline need to be negotiate again.

**Comment by zhoonit** [comment]
ACR has been issued. (internal)

**Comment by zhoonit** [comment]
Tizen 6.5 items are done, untouched item should go to Tizen 7.0 item instead
